File: AI/lab03/internal/dataset/dataset.go
package dataset

import (
	"ai_lab3/internal/training"
	"bufio"
	"encoding/csv"
	"io"
	"os"
	"strconv"
)

func toPair(in []string) (training.Pair, error) {
	res, err := strconv.ParseFloat(in[0], 64)
	if err != nil {
		return training.Pair{}, err
	}
	resEncoded := make([]float64, 3)
	resEncoded[int(res)-1] = 1
	var features []float64
	for i := 1; i < len(in); i++ {
		res, err = strconv.ParseFloat(in[i], 64)
		if err != nil {
			return training.Pair{}, err
		}
		features = append(features, res)
	}
	return training.Pair{
		Response: resEncoded,
		Input:    features,
	}, nil
}

func Load(path string) (training.Pairs, error) {
	f, err := os.Open(path)
	defer func(f *os.File) {
		_ = f.Close()
	}(f)
	if err != nil {
		return nil, err
	}
	r := csv.NewReader(bufio.NewReader(f))
	var pairs training.Pairs
	for {
		record, readErr := r.Read()
		if readErr == io.EOF {
			break
		}
		p, pairErr := toPair(record)
		if pairErr != nil {
			return nil, pairErr
		}
		pairs = append(pairs, p)
	}
	return pairs, nil
}

File: NN/lab01/src/dataset.py
import csv
from typing import List, Tuple

from src.training import Pair, Pairs

def to_pair(record: List[str]) -> Tuple[Pair]:
    res = int(float(record[0])) - 1
    res_encoded = [0.0] * 3
    res_encoded[res] = 1.0
    features = [float(x) for x in record[1:]]
    return Pair(input_data=features, response=res_encoded)

def load(path: str) -> Tuple[Pairs]:
    with open(path, mode='r', newline='') as f:
        reader = csv.reader(f)
        pairs = Pairs([])
        for record in reader:
            pair = to_pair(record)
            pairs.pairs.append(pair)
        return pairs
